File: app/src/main/java/ua/fvadevand/reminderstatusbar/dialogs/AlarmSetDialog.java
package ua.fvadevand.reminderstatusbar.dialogs;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Context;
import android.os.Bundle;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.DatePicker;
import android.widget.TextView;
import android.widget.TimePicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import java.util.Calendar;

import ua.fvadevand.reminderstatusbar.R;
import ua.fvadevand.reminderstatusbar.utilities.ReminderDateUtils;

public class AlarmSetDialog extends DialogFragment {

    public static final String TAG = "AlarmSetDialog";
    private static final String ARG_CALENDAR = "calendar";
    private OnAlarmSetListener mListener;
    private TextView mTimeTextView;
    private TextView mDateTextView;
    private Calendar mCalendar;

    TimePickerDialog.OnTimeSetListener timeCallBack = new TimePickerDialog.OnTimeSetListener() {
        @Override
        public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
            mCalendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            mCalendar.set(Calendar.MINUTE, minute);
            mTimeTextView.setText(ReminderDateUtils.formatTime(view.getContext(), mCalendar));
        }
    };

    DatePickerDialog.OnDateSetListener dateCallBack = new DatePickerDialog.OnDateSetListener() {
        @Override
        public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
            mCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
            mCalendar.set(Calendar.MONTH, month);
            mCalendar.set(Calendar.YEAR, year);
            mDateTextView.setText(ReminderDateUtils.formatFullDate(view.getContext(), mCalendar));
        }
    };

    public static AlarmSetDialog newInstance(Calendar calendar) {
        AlarmSetDialog fragment = new AlarmSetDialog();
        Bundle args = new Bundle();
        args.putSerializable(ARG_CALENDAR, calendar);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null)
            mCalendar = (Calendar) getArguments().getSerializable(ARG_CALENDAR);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Context context = getActivity();
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        @SuppressLint("InflateParams") View rootView = layoutInflater.inflate(R.layout.dialog_alarm, null);
        mTimeTextView = rootView.findViewById(R.id.tv_time);
        mTimeTextView.setText(ReminderDateUtils.formatTime(context, mCalendar));
        mTimeTextView.setOnClickListener(v -> {
            int hour = mCalendar.get(Calendar.HOUR_OF_DAY);
            int minute = mCalendar.get(Calendar.MINUTE);
            TimePickerDialog timeDialog = new TimePickerDialog(v.getContext(),
                    timeCallBack,
                    hour,
                    minute,
                    DateFormat.is24HourFormat(v.getContext()));
            timeDialog.show();
        });

        mDateTextView = rootView.findViewById(R.id.tv_date);
        mDateTextView.setText(ReminderDateUtils.formatFullDate(context, mCalendar));
        mDateTextView.setOnClickListener(v -> {
            int year = mCalendar.get(Calendar.YEAR);
            int month = mCalendar.get(Calendar.MONTH);
            int day = mCalendar.get(Calendar.DAY_OF_MONTH);
            DatePickerDialog dateDialog = new DatePickerDialog(v.getContext(),
                    dateCallBack,
                    year,
                    month,
                    day);
            dateDialog.show();
        });

        return new AlertDialog.Builder(context)
                .setView(rootView)
                .setTitle(R.string.alarm_dialog_title)
                .setPositiveButton(android.R.string.ok, (dialog, id) -> mListener.onAlarmSet(mCalendar))
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> {

                })
                .create();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mListener = (OnAlarmSetListener) getParentFragment();
        } catch (ClassCastException e) {
            throw new ClassCastException(getParentFragment() + " must implement OnAlarmSetListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnAlarmSetListener {
        void onAlarmSet(Calendar calendar);
    }
}

File: app/src/main/java/ua/fvadevand/reminderstatusbar/dialogs/AlarmSetDialog.kt
package ua.fvadevand.reminderstatusbar.dialogs

import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.app.Dialog
import android.app.TimePickerDialog
import android.content.Context
import android.os.Bundle
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import ua.fvadevand.reminderstatusbar.R
import ua.fvadevand.reminderstatusbar.adapters.PeriodTypesAdapter
import ua.fvadevand.reminderstatusbar.data.models.PeriodType
import ua.fvadevand.reminderstatusbar.data.models.PeriodType.PeriodTypes
import ua.fvadevand.reminderstatusbar.utils.ReminderDateUtils
import java.util.Calendar

class AlarmSetDialog : DialogFragment() {

    private lateinit var timeTextView: TextView
    private lateinit var dateTextView: TextView
    private lateinit var calendar: Calendar
    private var periodType = PeriodType.ONE_TIME
    private var listener: OnAlarmSetListener? = null

    private var timeCallBack: TimePickerDialog.OnTimeSetListener = TimePickerDialog.OnTimeSetListener { view, hourOfDay, minute ->
        with(calendar) {
            set(Calendar.HOUR_OF_DAY, hourOfDay)
            set(Calendar.MINUTE, minute)
        }
        timeTextView.text = ReminderDateUtils.formatTime(view.context, calendar)
    }

    private var dateCallBack: DatePickerDialog.OnDateSetListener = DatePickerDialog.OnDateSetListener { view, year, month, dayOfMonth ->
        with(calendar) {
            set(Calendar.DAY_OF_MONTH, dayOfMonth)
            set(Calendar.MONTH, month)
            set(Calendar.YEAR, year)
        }
        dateTextView.text = ReminderDateUtils.formatFullDate(view.context, calendar)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        calendar = Calendar.getInstance()
        calendar.timeInMillis = arguments?.getLong(ARG_CALENDAR) ?: System.currentTimeMillis()
        periodType = arguments?.getInt(ARG_PERIOD_TYPE) ?: PeriodType.ONE_TIME
    }

    @SuppressLint("InflateParams")
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val context = context!!
        val rootView = LayoutInflater.from(context).inflate(R.layout.dialog_alarm, null)
        timeTextView = rootView.findViewById(R.id.tv_alarm_dialog_time)
        timeTextView.text = ReminderDateUtils.formatTime(context, calendar)
        timeTextView.setOnClickListener { v ->
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            val minute = calendar.get(Calendar.MINUTE)
            val timeDialog = TimePickerDialog(v.context,
                    timeCallBack,
                    hour,
                    minute,
                    DateFormat.is24HourFormat(v.context))
            timeDialog.show()
        }

        dateTextView = rootView.findViewById(R.id.tv_alarm_dialog_date)
        dateTextView.text = ReminderDateUtils.formatFullDate(context, calendar)
        dateTextView.setOnClickListener { v ->
            val year = calendar.get(Calendar.YEAR)
            val month = calendar.get(Calendar.MONTH)
            val day = calendar.get(Calendar.DAY_OF_MONTH)
            val dateDialog = DatePickerDialog(v.context,
                    dateCallBack,
                    year,
                    month,
                    day)
            dateDialog.show()
        }
        val spinner: Spinner = rootView.findViewById(R.id.spinner_alarm_dialog_repeat)
        val periodTypes = PeriodType.getPeriodTypes()
        val adapter = PeriodTypesAdapter(context, periodTypes)
        spinner.adapter = adapter
        for (i in 0 until periodTypes.size) {
            if (periodType == periodTypes[i]) {
                spinner.setSelection(i)
                break
            }
        }
        return AlertDialog.Builder(context)
                .setView(rootView)
                .setTitle(R.string.alarm_dialog_title)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    listener?.onAlarmSet(calendar.timeInMillis, adapter.getItem(spinner.selectedItemPosition))
                }
                .setNegativeButton(android.R.string.cancel) { _, _ -> }
                .create()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        listener = parentFragment as OnAlarmSetListener?
    }

    override fun onDetach() {
        listener = null
        super.onDetach()
    }

    interface OnAlarmSetListener {
        fun onAlarmSet(alarmTimeInMillis: Long, @PeriodTypes periodType: Int)
    }

    companion object {

        const val TAG = "AlarmSetDialog"
        private const val ARG_CALENDAR = "calendar"
        private const val ARG_PERIOD_TYPE = "period_type"

        fun newInstance(timeInMillis: Long, @PeriodTypes periodType: Int): AlarmSetDialog {
            val fragment = AlarmSetDialog()
            val args = Bundle()
            args.putLong(ARG_CALENDAR, timeInMillis)
            args.putInt(ARG_PERIOD_TYPE, periodType)
            fragment.arguments = args
            return fragment
        }
    }
}
